lib/auth/saml.go

```diff
@@ -308,80 +346,159 @@ type SAMLAuthResponse struct {
 		}
-	}
+		event.Status.Success = true
+		event.User = re.auth.Username
+		event.Code = events.UserSSOLoginCode
 
-	if err != nil {
-		event.Code = events.UserSSOLoginFailureCode
-		event.Status.Success = false
-		event.Status.Error = trace.Unwrap(err).Error()
-		event.Status.UserMessage = err.Error()
 		if err := a.emitter.EmitAuditEvent(a.closeCtx, event); err != nil {
-			log.WithError(err).Warn("Failed to emit SAML login failed event.")
+			log.WithError(err).Warn("Failed to emit SAML login event.")
 		}
-		return nil, trace.Wrap(err)
 	}
-	event.Status.Success = true
-	event.User = re.auth.Username
-	event.Code = events.UserSSOLoginCode
 
-	if err := a.emitter.EmitAuditEvent(a.closeCtx, event); err != nil {
-		log.WithError(err).Warn("Failed to emit SAML login event.")
+	if err != nil {
+		return nil, di, err
+	} else {
+		return &re.auth, di, nil
 	}
-
-	return &re.auth, nil
 }
 
 type samlAuthResponse struct {
 	auth                SAMLAuthResponse
 	attributeStatements map[string][]string
 }
 
-func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, error) {
+type SsoDiagnosticInfo struct {
+	Success     bool
+	SSOTestFlow bool
+	DiagMap     map[string]interface{}
+	Err         string
+	// original request; one of:
+	SAMLRequest   *services.SAMLAuthRequest
+	OIDCRequest   *services.OIDCAuthRequest
+	GithubRequest *services.GithubAuthRequest
+}
+
+const (
+	DiagInfoProblem                 = "common.problem"
+	DiagInfoError                   = "common.error"
+	DiagInfoRequestId               = "common.requestID"
+	DiagInfoCreateUserParams        = "common.createUserParams"
+	DiagInfoSAMLAttributesToRoles   = "SAML.attributesToRoles"
+	DiagInfoSAMLAssertionInfo       = "SAML.assertionInfo"
+	DiagInfoSAMLAttributeStatements = "SAML.attributeStatements"
+)
+
+func (di *SsoDiagnosticInfo) GetClientRedirectUrl() (string, error) {
+	clientRedirectUrl := ""
+
+	if di != nil {
+		if di.SAMLRequest != nil {
+			clientRedirectUrl = di.SAMLRequest.ClientRedirectURL
+		} else if di.OIDCRequest != nil {
+			clientRedirectUrl = di.OIDCRequest.ClientRedirectURL
+		} else if di.GithubRequest != nil {
+			clientRedirectUrl = di.GithubRequest.ClientRedirectURL
+		}
+	}
+
+	if clientRedirectUrl == "" {
+		return "", trace.Errorf("unable to find client redirect URL.")
+	}
+
+	return clientRedirectUrl, nil
+}
+
+func (di *SsoDiagnosticInfo) SetProblem(problem string, err error) {
+	di.DiagMap[DiagInfoProblem] = problem
+	di.DiagMap[DiagInfoError] = err.Error()
+	di.Err = err.Error()
+}
+
+func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, *SsoDiagnosticInfo, error) {
 	ctx := context.TODO()
+	di := &SsoDiagnosticInfo{
+		Success:     false,
+		DiagMap:     map[string]interface{}{},
+		SSOTestFlow: false,
+	}
+
 	requestID, err := parseSAMLInResponseTo(samlResponse)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		di.SetProblem("Failed to parse SAML response", err)
+		return nil, di, trace.Wrap(err)
```

Comment:
Does the di needs to be returned with err != nil 

What about decoupling error message and successful fileds from `SsoDiagnosticInfo` 
and proagating error to the http caller by  `return nil, trace.Wrap(err).WithField(SSODiagnosticErrors, di)`